### PR TITLE
fix: parseJWT is now throwing AuthenticationError.invalidCredentials …

### PR DESCRIPTION
**Pull request by mihailstumkins**
…instead of JWTError.incorrectNumberOfSegments

### CODE REVIEW

File: Sources/JWTProvider/Request+JWT.swift [hunk 1]
```diff
@@ -36,6 +36,16 @@ extension Request {
...
+        // Check if JWT Token can be parsed
+        let separator: Byte = .period
+        let segments = bearer.string.components(
+            separatedBy: [separator].makeString()
+        )
+
+        guard segments.count == 3 else {
+            throw AuthenticationError.invalidCredentials
+        }
+
```
Comment (siemensikkema) [inline Sources/JWTProvider/Request+JWT.swift]: I'd prefer catching the error from JWT. The reason being that the segment count is an implementation detail that is already handled by JWT.

```swift
do {
    // Parse the bearer string into a JWT
    return try JWT(token: bearer.string)
} catch JWTError.incorrectNumberOfSegments {
    throw AuthenticationError.invalidCredentials
}
```

**Review by siemensikkema — CHANGES_REQUESTED** [review]
One comment on the redundant check for segments. Test looks good!

**Review by siemensikkema — APPROVED** [review]

**Comment by siemensikkema** [comment]
@tanner0101 if you approve as well, can we make this part of 1.2.0?

**Comment by tanner0101** [comment]
I'm concerned here that hiding this error could result in some confusion about _why_ invalid auth error is being thrown. We're throwing away the useful information that the JWT was invalid here. Unfortunately we can't modify the enum at all to include the error, so maybe the only solution is to print out a message using `print`.

Other options are create a new error for this situation, or conforming that particular enum case to AbortError: 401 (or 403 whichever it is)

**Comment by siemensikkema** [comment]
@tanner0101 Good point. Since conforming only one case to AbortError is not possible we'd have to find meaningful response codes for the other cases as well. I guess the easiest is to do a print statement but that is only helpful for developers with access to the logs.

**Comment by tanner0101** [comment]
Oops, that was a bit confusing what I wrote. By conform the one case I meant something like:

```swift
extension JWTError: AbortError {
    var status: Status { 
        switch self {
        case .invalidSegments: return .invalidAuth
        default: .internalServerError
        }
    }
}
```

I think printing is also an "ok" solution. We've used it in the past.

**Comment by siemensikkema** [comment]
The available status code would be `unauthorized`. Wouldn't treating all other JWTError cases as internal server error also be confusing? A lot of those are actually client side errors ... That's what I meant by finding meaningful status codes. We could also involve `Debuggable` to provide more info.
@mihailstumkins what do you think?

**Comment by mihailstumkins** [comment]
I agree - solution with JWTError extension to confirm to AbortError seems to be better.
if we are confirming to Debuggable we could use JWTError.description for reason as well, the only thing I'm about in this case is what should we use as value for identifier property. Maybe also description?

this is how JWTError description is created in JWTError.swift file

```
extension JWTError: CustomStringConvertible {
    public var description: String {
        let reason: String

        switch self {
        case .createKey:
            reason = "Could not create key"
        case .createPublicKey:
            reason = "Could not create public key"
        case .decoding:
            reason = "Could not decode"
        case .encoding:
            reason = "Could not encode"
        case .incorrectNumberOfSegments:
            reason = "Incorrect number of segments"
        case .incorrectPayloadForClaimVerification:
            reason = "Payload is not of type `[String: Node]`"
        case .missingAlgorithm:
            reason = "Missing algorithm"
        case .missingClaim(withName: let name):
            reason = "Missing claim with name: \(name)"
        case .privateKeyRequired:
            reason = "Private key required"
        case .signing:
            reason = "Could not sign"
        case .wrongAlgorithm:
            reason = "Wrong algorithm"
        case .verificationFailedForClaim(withName: let name):
            reason = "Claim verification failed for claim with name: \(name)"
        case .signatureVerificationFailed:
            reason = "Signature verification failed"
        case .unknown(let error):
            reason = "\(error)"
        }

        return "JWT error: \(reason)"
    }
}
```

so if I add JWTError+Status.swift with following content:
```
import Vapor
import HTTP
import JWT


extension JWTError: AbortError {
   public var status: Status {
        switch self {
        case .incorrectNumberOfSegments:
            return .unauthorized
        default:
            return .internalServerError
        }
    }
}

extension JWTError: Debuggable {
    public var reason: String {
        return self.description
    }
    
    public var identifier: String {
        return self.description
    }
    
    public var possibleCauses: [String] {
        return []
    }
    
    public var suggestedFixes: [String] {
        return []
    }
}
```

I would get following result http://prntscr.com/gz9gpf
@tanner0101 @siemensikkema  what do you guys think?

**Comment by tanner0101** [comment]
@mihailstumkins that seems like the best solution 👍

**Comment by vzsg** [comment]
I would also include incorrectPayloadForClaimVerification, missingAlgorithm, missingClaim, wrongAlgorithm, verificationFailedForClaim and signatureVerificationFailed to the unauthorized group, as they all indicate an invalid token being presented by the client.

(Disregard this if it was implied, just not added to keep the example clear.)

**Comment by mihailstumkins** [comment]
added changes as discussed. @tanner0101 , @siemensikkema , @vzsg

**Comment by siemensikkema** [comment]
@tanner0101 see above circle ci issues. I don't have the privileges to attempt to fix this. Do you know what could be wrong?

**Comment by mihailstumkins** [comment]
After the merge with maser via 'Update branch' button, circle ci issue seems to resolve. It's weird since there are no actual code changes with this merge.

**Comment by mihailstumkins** [comment]
So guys what are the roadmap for including this into a release? Is something else is needed from my side? @tanner0101 @siemensikkema @vzsg

**Review by tanner0101 — APPROVED** [review]

**Comment by tanner0101** [comment]
LGTM, thanks @mihailstumkins

**Comment by tanner0101** [comment]
https://github.com/vapor/jwt-provider/releases/tag/1.3.0

**Pull request merged**